File: content/snippets/go/interview_lcci_10_10_rank_from_stream.go
package main

import "fmt"

type binaryIndexedTree struct {
	tree []int
}

func newBinaryIndexedTree(n int) *binaryIndexedTree {
	return &binaryIndexedTree{tree: make([]int, n+1)}
}

func (b *binaryIndexedTree) update(index int, delta int) {
	for index < len(b.tree) {
		b.tree[index] += delta
		index += index & -index
	}
}

func (b *binaryIndexedTree) query(index int) int {
	sum := 0
	for index > 0 {
		sum += b.tree[index]
		index -= index & -index
	}
	return sum
}

type StreamRank struct {
	bit *binaryIndexedTree
}

func Constructor() StreamRank {
	return StreamRank{bit: newBinaryIndexedTree(50010)}
}

func (s *StreamRank) Track(x int) {
	s.bit.update(x+1, 1)
}

func (s *StreamRank) GetRankOfNumber(x int) int {
	return s.bit.query(x + 1)
}

func main() {
	streamRank := Constructor()
	fmt.Println(streamRank.GetRankOfNumber(1))
	streamRank.Track(0)
	fmt.Println(streamRank.GetRankOfNumber(0))
}

File: content/snippets/cpp/interview_lcci_10_10_rank_from_stream.cpp
#include <iostream>
#include <vector>
using namespace std;

class BinaryIndexedTree {
public:
    explicit BinaryIndexedTree(int n) : tree(n + 1, 0) {}

    void update(int index, int delta) {
        while (index < static_cast<int>(tree.size())) {
            tree[index] += delta;
            index += index & -index;
        }
    }

    int query(int index) const {
        int sum = 0;
        while (index > 0) {
            sum += tree[index];
            index -= index & -index;
        }
        return sum;
    }

private:
    vector<int> tree;
};

class StreamRank {
public:
    StreamRank() : bit(50010) {}

    void track(int x) {
        bit.update(x + 1, 1);
    }

    int getRankOfNumber(int x) const {
        return bit.query(x + 1);
    }

private:
    BinaryIndexedTree bit;
};

int main() {
    StreamRank streamRank;
    cout << streamRank.getRankOfNumber(1) << endl;
    streamRank.track(0);
    cout << streamRank.getRankOfNumber(0) << endl;
    return 0;
}
